File: rust/advent/src/bin/2023/y23d08.rs
use rayon::iter::ParallelIterator;
use std::collections::HashMap;
use rayon::iter::ParallelBridge;
use advent::utilities::get_input::get_input;
use utilities::{
    math::formulae::lcm,
    structs::stopwatch::{ReportDuration, Stopwatch},
};

type Input<'a> = (&'a str, HashMap<&'a str, (&'a str, &'a str)>);
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 8).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input<'_> {
    let (directions, net_str) = input.split_once("\n\n").unwrap();
    let mut network= HashMap::new();
    for line in net_str.lines() {
        let (node, to) = line.split_once(" = (").unwrap();
        let (left, to) = to.split_once(", ").unwrap();
        let (right, _) = to.split_once(')').unwrap();
        network.insert(node, (left, right));
    }
    (directions, network)
}

fn traverse<F>(
    directions: &str,
    network: &HashMap<&str, (&str, &str)>,
    start_node: &str,
    end_condition: F,
) -> usize
where
    F: Fn(&str) -> bool,
{
    directions
        .chars()
        .cycle()
        .scan(start_node, |node, dir| {
            let new_node = network
                .get(node)
                .map(|&(left, right)| {
                    if dir == 'L' {
                        left
                    } else {
                        right
                    }
                })
                .unwrap();
            *node = new_node;
            Some(new_node)
        })
        .enumerate()
        .find(|(_, node)| end_condition(node))
        .unwrap()
        .0
        + 1
}

fn part1(input: &Input) -> Output {
    let (directions, network) = input;
    let end_condition = |end: &str| end == "ZZZ";
    traverse(directions, network, "AAA", end_condition)
}

fn part2(input: &Input) -> Output {
    let (directions, network) = input;
    let end_condition = |end: &str| end.ends_with('Z');
    network
        .keys()
        .filter(|node| node.ends_with('A'))
        .par_bridge()
        .map(|node| traverse(directions, network, node, end_condition))
        .reduce(|| 1, |acc, cycle_length| lcm(acc as i64, cycle_length as i64) as usize)
}

#[test]
fn default() {
    let input = get_input(23, 8).unwrap();
    let input = parse_input(&input);
    assert_eq!(19241, part1(&input));
    assert_eq!(9606140307013, part2(&input));
}

// Input parsed (151μs)
// 1. 19241 (429μs)
// 2. 9606140307013 (1ms)
// Total: 1ms

File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D08.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.math.lcm
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit

class Y23D08(input: String) : Day {

    private val directions: Sequence<Char>
    private val network: Map<String, Pair<String, String>>
    
    init {
        val (dirStr, netStr) = input.blankSplit()
        
        // infinite sequence of directions (repeats after string ends)
        directions = generateSequence(dirStr.asSequence()) { it }.flatten()
        
        // representing a puzzle map with a Kotlin map
        network = netStr
            .lines()
            .associate { line -> 
                val (node, left, right) = line.split(" = (", ", ", ")")
                node to Pair(left, right)
            }
    }
        
    // deliver steps needed for traveler to go from startNode to an end condition,
    private inline fun traverse(startNode: String, endCondition: (String) -> Boolean): Int = directions
        .runningFold(startNode) { node, dir ->
            network.getValue(node).let { (left, right) -> if (dir == 'L') left else right }
        }.indexOfFirst(endCondition)
    
    override fun part1() = traverse("AAA") { it == "ZZZ" }

    // find how long it takes for each ghost to make it from a to z, making the *massive* assumption that
    // they immediately cycle (they do). These are all different. So lcm gives you the answer.
    override fun part2(): Long = network.keys
        .filter { it.last() == 'A' }
        .map { node -> traverse(node) { it.last() == 'Z' }.toLong() }
        .lcm()
}

fun main() = Day.runDay(Y23D08::class)

//    Class creation: 16ms
//    Part 1: 19241 (13ms)
//    Part 2: 9606140307013 (31ms)
//    Total time: 61ms

@Suppress("unused")
private val sampleInput = listOf(
    """RL

AAA = (BBB, CCC)
BBB = (DDD, EEE)
CCC = (ZZZ, GGG)
DDD = (DDD, DDD)
EEE = (EEE, EEE)
GGG = (GGG, GGG)
ZZZ = (ZZZ, ZZZ)
""", """LLR

AAA = (BBB, BBB)
BBB = (AAA, ZZZ)
ZZZ = (ZZZ, ZZZ)
""",
)

@Suppress("unused")
private val sampleInput2 = listOf(
    """LR

11A = (11B, XXX)
11B = (XXX, 11Z)
11Z = (11B, XXX)
22A = (22B, XXX)
22B = (22C, 22C)
22C = (22Z, 22Z)
22Z = (22B, 22B)
XXX = (XXX, XXX)
""" to "6"
)